Cryptenroll specify handle index

ddstreet opened the pull request:
Allow user to specify what handle index to use for systemd-cryptenroll to use for tpm2 sealing. This defaults to the SRK.

This is on top of #29426 and split out from #28519.

bluca commented on man/systemd-cryptenroll.xml:
```diff
@@ -411,6 +411,26 @@
+
+        <listitem><para>Configures which parent key to use for sealing, using the TPM handle of the
+        key. Expects a hexadecimal 32bit integer, optionally prefixed with <literal>0x</literal>. Allowable
+        values are any handle index in the persistent (0x81000000-0x81ffffff), transient
```
wrap these ranges in `<literal>` too

ddstreet replied:
ack, updated, thanks

bluca reviewed (changes requested):
Can you add a test for this? Doesn't have to be complex, just the minimum to send this through ASAN&friends

poettering commented on man/systemd-cryptenroll.xml:
```diff
@@ -411,6 +411,26 @@
       </varlistentry>
 
+      <varlistentry>
+        <term><option>--tpm2-handle=</option><replaceable>HANDLE</replaceable></term>
```
can we call this `--tpm2-storage-key=` maybe? i.e. "handle" is just so generic, I think it would be better to say that we reference a *key* here, and what kind of key it is. The handle after all is not the primary concept we want to configure here, but the the object the handle refers to...

other ideas: `--tpm2-parent-key=` or `--tpm2-primary-key=` or `--tpm2-root-key=` or `--tpm2-wrap-key=` or `--tpm2-seal-key=` something like that.

poettering replied:
so i read from the other PR that this is supposed to deal with a numeric handler as opposed to a key file, hence maybe call this `--tpm2-seal-key-handle=` and `--tpm2-seal-key-file=`?

ddstreet replied:
sounds good, updated to `--tpm2-seal-key-handle=`

poettering commented on src/cryptenroll/cryptenroll-tpm2.c:
```diff
@@ -133,6 +133,7 @@ int enroll_tpm2(struct crypt_device *cd,
                 const void *volume_key,
                 size_t volume_key_size,
                 const char *device,
+                uint32_t handle,
```
i'd also call this `seal_key` or `storage_key` or so (depending on how the cmdline switch is going to be called)

ddstreet replied:
ack, updated to `seal_key_handle` (since we'll add `seal_key_file` later also)

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -3931,6 +3932,7 @@ static int tpm2_deserialize(
 }
 
 int tpm2_seal(Tpm2Context *c,
+              uint32_t handle_index,
```
so here you call it the "handle_index", and earlier you called it the "handle". Please stick to one. (And as mentioned, I think we should rather call this storage_key or sealing_key or so)

ddstreet replied:
ack, changed it to `seal_key_handle` here as well.

poettering commented:
this should have a test, really.

ddstreet commented:
Updated to address review comments, as well as adding tests (to both test-tpm2 and TEST-70-TPM2).

I also looked more closely at the spec and realized NV indexes cannot be used as keys, so I removed NV index from the allowable ranges for `--tpm2-seal-key-handle=`; now it must be either a persistent key or transient key.

Also the test uncovered that we can't use Esys_TR_Close(), so added a commit to stop calling it. Technically if tpm2-tss library added internal reference counting to its handles (so Esys_TR_Close() didn't remove a handle in use by multiple callers) we could call it again, but for now it's easiest (and I think ok) to just avoid calling it. I also opened a [tpm2-tss issue](https://github.com/tpm2-software/tpm2-tss/issues/2693).

ddstreet commented on test/units/testsuite-70.sh:
```diff
@@ -131,6 +138,57 @@
+tpm2_flushcontext -t
+
+systemd-cryptenroll --wipe-slot=tpm2 "$img"
+PASSWORD=passphrase systemd-cryptenroll --tpm2-device=auto --tpm2-seal-key-handle=${persistent#0x} "$img"
```
ok

ddstreet commented on test/units/testsuite-70.sh:
```diff
@@ -131,6 +138,57 @@
+systemd-cryptsetup detach test-volume
+
+systemd-cryptenroll --wipe-slot=tpm2 "$img"
+PASSWORD=passphrase systemd-cryptenroll --tpm2-device=auto --tpm2-seal-key-handle=$persistent "$img"
```
ack

poettering commented on src/shared/tpm2-util.c:
```diff
@@ -699,12 +699,22 @@ static void tpm2_handle_cleanup(ESYS_CONTEXT *esys_context, ESYS_TR esys_handle,
+                 * references will be cleaned up when we free our ESYS_CONTEXT.
+                 *
+                 * An upstream bug is open here: https://github.com/tpm2-software/tpm2-tss/issues/2693 */
+                rc = TSS2_RC_SUCCESS /* sym_Esys_TR_Close(esys_context, &esys_handle) */;
```
maybe add a `//` FIXME style comment here

i.e. `// FIXME: restore sym_Esys_TR_Close() once tpm2-tss is fixed and adopted widely enough`

ddstreet replied:
ack, updated

poettering commented:
lgtm

ddstreet commented:
rebased on main, updated with // FIXME comment, and fixed test/units/testsuite-70.sh cleanup function to use if-fi instead of && (so it doesn't incorrectly cause test to fail)

ddstreet commented:
Hmm, on Arch it fails to find the transient key, it seems like GetCapability for transient handles might not work there...I'll boot up my Arch VM and take a look

ddstreet commented:
Added one more commit to avoid calling GetCapability() to check if a handle exists in the tpm, if the handle is transient, because the kernel tpm "resource manager" (i.e. /dev/tpmrm0) will never report that any transient handles exist, even if they actually do.
